### PR TITLE
fix(compat): cjs/esm interop for dynamic imports

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit fixes CJS/ESM interop in compat mode for dynamically
imported modules.

"ProcState::prepare_module_load" was changed to accept a list
of "graph roots" without associated "module kind". That module kind
was always hardcoded to "ESM" which is not true for CJS/ESM interop -
a CommonJs module might be imported using "import()" function. In
such case the root of the graph should have "CommonJs" module kind
instead of "ESM".

Fixes https://github.com/denoland/deno/issues/13790

### CODE REVIEW

**Comment by bartlomieju** [comment]
Culprit is here:
https://github.com/denoland/deno/blob/a41d399f5f728c73b652c3f5f5c2bf57d2054153/cli/module_loader.rs#L109-L120

Root of the graph is always marked as ESM, which in this instance is not correct.

**Comment by bartlomieju** [comment]
@kitsonk could you take a look? I feel we should be using resolver inside `ModuleLoader::prepare_load` and pass resolved specifier and module kind based on the `ResolveResponse` instead of hard coding module kind to `Esm`.

**Comment by bartlomieju** [comment]
ping @kitsonk

File: cli/proc_state.rs [hunk 1]
```diff
@@ -252,13 +253,46 @@ impl ProcState {
...
+    // One might argue that this is a code smell, and I would agree. However
+    // due to flux in "Node compatibility" it's not clear where it should be
+    // decided what `ModuleKind` is decided for root specifier.
+    let roots = roots
+      .into_iter()
+      .map(|r| {
+        if let Some(resolver) = &maybe_resolver {
+          let response =
+            resolver.resolve(r.as_str(), &Url::parse("unused:").unwrap());
+          // TODO(bartlomieju): this should be implemented in `deno_graph`
```
Comment (kitsonk) [inline cli/proc_state.rs:284]: Ref: denoland/deno_graph#140

**Review by kitsonk — APPROVED** [review]
LGTM

File: cli/proc_state.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/proc_state.rs:284]: Thanks for opening this!

**Pull request merged**